paper.tex

```diff
@@ -107,20 +108,48 @@ \section{Trading Model}
+
+\begin{figure}[h]
+	\centering
+	\includegraphics[width=0.5\textwidth]{key_certification}
+	\caption{Session key certification process with blind signature.}
+	\label{fig:key_certification}
+\end{figure}
 
 The contract address and product description will be stored in a file which is uploaded to IPFS. Consumers can search the desired product by keywords or tags. The consumer then evaluates the product and start trading with the data provider if the consumer is interested in subscribing the data.
 
 \subsection{Trading}
-Once a consumer who want to subscribe the sensor data pays a subscription fee to the Product Contract, it is added to the subscriber list automatically by the smart contract. To decrypt the streaming data on the MAM channel, session key should be exchanged between data provider and subscribers. The data provider requests the subscription list maintained by the Product Contract and obtains accounts of each subscriber. After that, the data provider can also get public keys of each subscriber from the Registration Contact with their accounts. For each subscriber, the data provider encrypts the session key and broker's signature with the subscriber's public key and sends the ciphertext to the Product Contract. Subscribers listen to the smart contract event which is triggered when the ciphertext is updated, and decrypt the ciphertext to get the session key and signature. Subscribers can obtain the broker's public key on the Registration Contact as well, so they can verify that the signature is valid and the session key is the only one that is certified by the broker. Afterward, encrypted sensing data is uploaded to the MAM channel, and subscribers can download and decrypt them with the session¸
+Once a consumer who want to subscribe the sensor data pays a subscription fee to the Product Contract, it is added to the consumer list automatically by the smart contract. To decrypt the streaming data on the MAM channel, session key should be exchanged between data provider and consumers as shown in Figure \ref{fig:key_exchange}. The data provider requests the subscription list maintained by the Product Contract and obtains accounts of each consumer. After that, the data provider can also get public keys of each consumer from the Registration Contact with their accounts. For each consumer, the data provider encrypts the session key and broker's signature with the consumer's public key and sends the ciphertext to the Product Contract. Consumers listen to the smart contract event which is triggered when the ciphertext is updated, and decrypt the ciphertext to get the session key and signature. Consumers can obtain the broker's public key on the Registration Contact as well, so they can verify that the signature is valid and the session key is the only one that is certified by the broker. Afterward, encrypted sensing data is uploaded to the MAM channel, and consumers can download and decrypt them with the session¸
```

Comment:
The last sentence is not finished, `and decrypt them with the session key.`, `key` is missing here.